### DISCUSSION TITLE
Migration Puppet 4 > 7

### DISCUSSION THREAD

**Discussion by valentinunistra** [puppetlabs/community#60] [General]
Hello everyone,

If i want migrate Puppet from Puppet version 4 to Puppet version 7, what is the mains keys/steps to look ? I want to maintain my configuration.

**Comment by bastelfreak** [comment]
Hi,
there are a few options and constraints.
* I assume you haven't updated your modules in ages because new ones don't work with such an old Puppet version. Even puppet 7 is soon EoL, why not migrate to Puppet 8?
* Which OSes do you use? Puppet 7/8 don't support ancient versions and on the other side Puppet 4 doesn't run properly on modern distributions. So do you need to do some dist-upgrades as well?

There changed so much in the ecosystem and I assume you've zero tests. I predict: rewriting everything and building a proper CI pipeline is quicker than updating the existing code base. But to make proper estimates you need to provide more information

* How many nodes
* Which operating systems
* How many internal/private modules, how many external modules?
* What kind of stuff do you automate on the systems?
* Is that a corporate environment / how much downtime is allowed?

**Comment by binford2k** [comment]
I'd suggest duplicating your environment onto a new Puppet 8 test infrastructure. Then use https://github.com/voxpupuli/onceover to smoke test the environment to see what breaks and https://github.com/voxpupuli/puppet-catalog_diff to see what changes in the catalog.

https://github.com/github/octocatalog-diff is a bit harder to set up, but if you have a large infrastructure it might be worth it.

If you have your own modules, we've just published a guide to updating them to Puppet 8. Keep in mind, this is intended for far more current codebases, so your milage may vary. https://dev.to/puppet/updating-modules-to-puppet-8-58h3
